internal/cache/resource_map.go

```diff
@@ -1,153 +1,126 @@
 package internal
 
 import (
+	"hash/maphash"
+	"iter"
+	"runtime"
 	"sync"
+	"sync/atomic"
 )
 
-// The resourceMapEntry is the value type in a ResourceMap. Access to an entry is guarded by lock and isDeleted.
+// ResourceMap is a concurrency-safe map. It deliberately does not expose bare Get or Put methods as
+// its concurrency model is based on the assumption that access to the backing values must be
+// strictly synchronized. Instead, all operations should be executed through the various Compute
+// methods.
 //
-// The reason ResourceMap does not directly insert the value type into the backing sync.Map is because of the
-// following scenario:
-// Routine 1 wants to delete an entry because it is not needed anymore, while routine 2 wants to execute an operation on
-// the entry. Routine 2 loads the value from the map and runs its operation. In the meantime, routine 1 simply deletes
-// the entry from the map, completely voiding routine 2's work. To prevent this, the additional resourceMapEntry layer
-// provides a sync.RWMutex and a flag indicating the entry has been deleted from the map (isDeleted). All operations
-// acquire the read lock and check isDeleted, bailing if the entry has been deleted between when it was loaded from the
-// map and the read lock was acquired. Conversely, the write lock on the entry must be held during deletion, and
-// isDelete must be set once the entry has been deleted and while still holding the write lock. This way, only one
-// routine can successfully delete the entry from the map, but it must wait for all other in-flight read operations to
-// do so (see ResourceMap.DeleteIf for additional info).
-type resourceMapEntry[T any] struct {
-	// The write lock must be held while inserting or removing the entry from the resources map. The read lock must be
-	// held while interacting with value. Before interacting with the value, isDeleted must be checked.
-	lock sync.RWMutex
-	// isDeleted must always be checked before accessing value. If true, this entry should be entirely discarded.
-	// Guarded by lock.
-	isDeleted bool
-	// the actual value
-	value T
+// It uses striped locking as the synchronization mechanism. An array of locks is initialized, and
+// the key is hashed then modded to pick the corresponding lock. The compute operations will hold
+// the lock for the entire duration of the operation.
+//
+// The number of locks is picked using a sensible value, in this case the value returned by
+// [runtime.GOMAXPROCS], which represents the maximum number of CPUs that will be executing
+// simultaneously.
+type ResourceMap[K comparable, V any] struct {
+	locks   []sync.Mutex
+	stripes uint64
+	// TODO: This map can likely be replaced with [github.com/puzpuzpuz/xsync.MapOf], which offers a
+	//  type-safe functional equivalent of sync.Map. However, until
+	//  https://github.com/puzpuzpuz/xsync/issues/160 is resolved, each Compute operation creates too much
+	//  garbage.
```

Comment:
In the PR description, it sounds like we adopted the xsync Map already, but here it's just a TODO. Could you update the PR description to clarify? Also, the PR name be more descriptive? like "Un-invent the wheel with Sync.Map", or "Un-invent custom resourceMapEntry"?

Reply:
Lol yeah I can change the PR description. I think I was just very happy to axe this gross code.

As for the version you were looking at, the issue I linked has been resolved, and the author of the repo will make the changes I suggested! So this will be much nicer.